store/store.go

```diff
@@ -244,82 +249,187 @@ func (bs *BlockStore) LoadSeenCommit(height int64) *types.Commit {
+	pruned, err := bs.pruneBlockMetaAndHashes(height)
+	if err != nil {
+		return pruned, err
 	}
 
-	pruned := uint64(0)
+	if err := bs.batchDelete(firstBlockPartKey, height); err != nil {
+		return pruned, err
+	}
+
+	if err := bs.batchDelete(blockCommitKey, height); err != nil {
+		return pruned, err
+	}
+
+	if err := bs.batchDelete(seenCommitKey, height); err != nil {
+		return pruned, err
+	}
+
+	return pruned, nil
+}
+
+// pruneBlockMetaAndHashes prunes all block metas from the base height to the retainHeight
+// but before each, loads the hash from the block meta to also prune from the block hash table
+func (bs *BlockStore) pruneBlockMetaAndHashes(retainHeight int64) (uint64, error) {
+	iter, err := bs.db.Iterator(
+		blockMetaKey(1),
+		blockMetaKey(retainHeight),
+	)
+	if err != nil {
+		panic(err)
+	}
+	defer iter.Close()
+
 	batch := bs.db.NewBatch()
 	defer batch.Close()
-	flush := func(batch dbm.Batch, base int64) error {
-		// We can't trust batches to be atomic, so update base first to make sure noone
-		// tries to access missing blocks.
-		bs.mtx.Lock()
-		bs.base = base
-		bs.mtx.Unlock()
-		bs.saveState()
-
-		err := batch.WriteSync()
-		if err != nil {
-			return fmt.Errorf("failed to prune up to height %v: %w", base, err)
-		}
-		batch.Close()
-		return nil
-	}
 
-	for h := base; h < height; h++ {
-		meta := bs.LoadBlockMeta(h)
-		if meta == nil { // assume already deleted
-			continue
-		}
-		if err := batch.Delete(calcBlockMetaKey(h)); err != nil {
-			return 0, err
+	pruned := uint64(0)
+	for iter.Valid() {
+		// delete the block meta
+		if err := batch.Delete(iter.Key()); err != nil {
+			return 0, fmt.Errorf("pruning error at height %d: %w", decodeHeightFromKey(iter.Key()), err)
 		}
-		if err := batch.Delete(calcBlockHashKey(meta.BlockID.Hash)); err != nil {
-			return 0, err
+
+		// load the block meta so we can also remove the block hash from the db
+		var pbbm = new(tmproto.BlockMeta)
+		bz := iter.Value()
+
+		err = proto.Unmarshal(bz, pbbm)
+		if err != nil {
+			panic(fmt.Errorf("unmarshal to tmproto.BlockMeta: %w", err))
 		}
-		if err := batch.Delete(calcBlockCommitKey(h)); err != nil {
-			return 0, err
+
+		blockMeta, err := types.BlockMetaFromProto(pbbm)
+		if err != nil {
+			panic(fmt.Errorf("error from proto blockMeta: %w", err))
 		}
-		if err := batch.Delete(calcSeenCommitKey(h)); err != nil {
-			return 0, err
+
+		if err := batch.Delete(blockHashKey(blockMeta.BlockID.Hash)); err != nil {
+			return 0, fmt.Errorf("pruning error at height %d: %w", decodeHeightFromKey(iter.Key()), err)
 		}
-		for p := 0; p < int(meta.BlockID.PartSetHeader.Total); p++ {
-			if err := batch.Delete(calcBlockPartKey(h, p)); err != nil {
-				return 0, err
+
+		pruned++
+		// avoid batches growing too large by flushing to database regularly
+		if pruned%1000 == 0 {
+			if err := iter.Error(); err != nil {
+				return pruned, err
+			}
+			if err := iter.Close(); err != nil {
+				return pruned, err
```

Comment:
We don't need to return `pruned` on error I think, but if we should then let's make sure it's correct and only update it when we actually write the batch -- as it is, it returns what we were _planning_ to prune, but not what we've actually deleted from the database.

Reply:
Yeah I think you're right in that we don't need to return it on error. I am not exactly sure why we return it in the first place. If pruning was successful then it will log the height that the block store pruned to. If it failed at some point it should log the height with which it failed.

Reply:
It's just for stats. If we prune to height 100,000 we may have deleted 99,999 blocks, or we may have deleted 1 -- it depends on what the base was before we pruned. This info is helpful to node operators if they e.g. wonder why the node suddenly halted for 30 seconds (hint: it pruned 100,000 blocks).